services/src/main/java/org/keycloak/services/clientpolicy/executor/ClientPolicyExecutor.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.services.clientpolicy.executor;
+
+import javax.ws.rs.core.MultivaluedMap;
+
+import org.keycloak.models.ClientModel;
+import org.keycloak.protocol.oidc.endpoints.request.AuthorizationEndpointRequest;
+import org.keycloak.protocol.oidc.utils.OAuth2CodeParser;
+import org.keycloak.protocol.oidc.utils.OIDCResponseType;
+import org.keycloak.provider.Provider;
+import org.keycloak.representations.idm.ClientRepresentation;
+import org.keycloak.services.clientpolicy.ClientPolicyException;
+import org.keycloak.services.clientpolicy.ClientPolicyEvent;
+import org.keycloak.services.clientregistration.ClientRegistrationContext;
+import org.keycloak.services.clientregistration.policy.RegistrationAuth;
+import org.keycloak.services.resources.admin.AdminAuth;
+
+/**
+ * This executor specifies what action is executed on the client to which to which {@link ClientPolicyProvider} is adopted.
+ * The executor can be executed on the events defined in {@link ClientPolicyEvent}.
+ * It is sufficient for the implementer of this executor to implement methods in which they are interested
+ * and {@link isEvaluatedOnEvent} method.
+ */
+public interface ClientPolicyExecutor extends Provider {
+
+    @Override
+    default void close() {
+    }
+
+    /**
+     * returns true if this executor is executed against the client.
+     * A executor can be implemented to be executed on some events while not on others.
+     * On the event specified as the parameter, this executor is skipped if this method returns false.
+     *
+     * @param event defined in {@link ClientPolicyEvent}
+     * @return true if this executor is executed on the event.
+     */
+    default boolean isExecutedOnEvent(String event) {return true;}
+
+    /**
+     * execute actions against the client
+     * on Dynamic Registration Endpoint access for creating client.
+     *
+     * @param context
+     * @param authType
+     * @throws {@link ClientPolicyException} - if something wrong happens when execution actions
+     */
+    default void executeOnDynamicClientRegister(
+            ClientRegistrationContext context,
+            RegistrationAuth authType)  throws ClientPolicyException {}
```

Comment:
My intention for throwing `ClientPolicyException` from `executeOn...()` methods is as follows :

* There is the chance that an executor refuses a request from a client. For example, an executor for enforcing PKCE. When a client sends an authorization request without `code_challenge` on Authorization Endpoint, this executor refuses this request

* If an executor refuses a request from a client, the executor needs to convey this situation to these methods' caller in order for them to send an appropriate error to the client. For the example above, `AuthorizationEndpoint` returns to the client 400 Bad Request with `error="invalid_client"`.

* To convey such the situation, I've used an exception.

Reply:
As the same as condition, I'll implement `default void executeOnClientUpdate(ClientUpdateContext context)`  
 instead of current several method on each events (Dynamic Client Registration, Dynamic Client Update, Admin REST API Client Registration, Admin REST API Client Update).

Reply:
Ignore the comment on the default implementation - I completely miss-read the code, somehow I read "throw new ClientPolicyException()" as the default implementation, not that the method "throws ClientPolicyException"

Having the single method for an events (instead isExecuted on and executeOn) will be good and I think it will clean up the PR a lot, as well as require less "boilerplate" to implement executors and conditions.

Reply:
I see, as you pointed out, I'll integrate these two method (`isExecuted()` and `executeOn...()`) onto one method (`executeOn...()`).